rclcpp/include/rclcpp/executor.hpp

```diff
@@ -368,21 +360,25 @@ class Executor
     // Put the executor's guard condition in
     guard_condition_handles.guard_conditions[1] = \
       interrupt_guard_condition_->data;
-    // Then fill the SubscriberHandles with ready subscriptions
-    size_t guard_cond_handle_index = start_of_timer_guard_conds;
-    for (auto & timer : timers) {
-      guard_condition_handles.guard_conditions[guard_cond_handle_index] = \
-        timer->guard_condition_->data;
-      guard_cond_handle_index += 1;
-    }
 
     rmw_time_t * wait_timeout = NULL;
     rmw_time_t rmw_timeout;
 
-    if (timeout >= std::chrono::duration<int64_t, T>::zero()) {
+    auto next_timer_duration = get_earliest_timer();
+    /* If the next timer timeout must preempt the requested timeout
+       or if the requested timeout blocks forever, and there exists a valid timer,
+       replace the requested timeout with the next timeout. */
+    if ( (next_timer_duration < timeout || timeout <
+      std::chrono::duration<int64_t,
+      T>::zero()) && next_timer_duration >= std::chrono::nanoseconds::zero())
```

Comment:
Uncrustify is always screwing up my ridiculous line lengths. I think I may instead put one or more of these clauses into bools, which makes this conditional more readable and may eliminate the whitespace problem...